sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -117,9 +122,11 @@ class SparkSession private(
   @InterfaceStability.Unstable
   @transient
   lazy val sessionState: SessionState = {
-    SparkSession.reflect[SessionState, SparkSession](
-      SparkSession.sessionStateClassName(sparkContext.conf),
-      self)
+    parentSessionState
+      .map(_.copy(this))
```

Comment:
Nice catch! Now enforcing initializing of `SessionState` in cloned `SparkSession` to force the copy at clone creation time.